digdag-core/src/main/java/io/digdag/core/database/BasicDatabaseStoreManager.java

```diff
@@ -224,45 +230,48 @@ public int getIndex()
     @SuppressFBWarnings("RCN_REDUNDANT_NULLCHECK_WOULD_HAVE_BEEN_A_NPE")
     public <T> T transaction(TransactionAction<T, D> action)
     {
-        try {
-            return transactionRetryExecutor.runInterruptible(() -> {
-                TransactionState ts = new TransactionState();
-                while (true) {
+        Handle nested = currentTransaction.get();
+        if (nested != null) {
+            return nested.inTransaction((h, status) -> action.call(h, h.attach(daoIface)));
+        }
+        else {
+            try {
+                return transactionRetryExecutor.runInterruptible(() -> {
                     try (Handle handle = dbi.open()) {
-                        T retval = handle.inTransaction((h, status) -> action.call(h, h.attach(daoIface), ts));
-                        if (ts.retryNext) {
-                            ts.retryNext = false;
-                            ts.retryCount += 1;
-                        }
-                        else {
-                            return retval;
-                        }
+                        currentTransaction.set(handle);
+                        return handle.inTransaction((h, status) -> action.call(h, h.attach(daoIface)));
                     }
-                }
-            });
-        }
-        catch (RetryGiveupException ex) {
-            Throwable innerException = ex.getCause();
-            throw Throwables.propagate(innerException);
-        }
-        catch (InterruptedException ex) {
-            throw Throwables.propagate(ex);
+                    finally {
+                        currentTransaction.set(null);
+                    }
+                });
+            }
+            catch (RetryGiveupException ex) {
+                Throwable innerException = ex.getCause();
+                throw Throwables.propagate(innerException);
+            }
+            catch (InterruptedException ex) {
+                throw Throwables.propagate(ex);
```

Comment:
If the interrupted state will be restored where this RuntimeException is caught, it's okay. If not, restoring the state here would be better.